# RAG SYSTEM

This is complete system of building RAG and testing along, after the RAG system we will be adding the LLM for retreival of the iinformation in a better Format*

INSTALL Libraries for pdf

In [14]:
!pip install langchain
!pip install pypdf
!pip install sentence-transformers

Here We have Uploaded the Pdf which we have to use for chunking and creating RAG System

In [3]:
!pip install langchain langchain-community pypdf
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader("CON.pdf")
docs = loader.load()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.6/346.6 kB 24.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 56.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.1/73.1 kB 5.1 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.32.4, but you have requests 2.34.2 which is incompatible.


/tmp/ipykernel_8341/1660283388.py:2: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders import PyPDFLoader


In [ ]:
Inspecting the pdf

In [23]:
print(docs[0].page_content[:1000])

THE BHARA TIY A NY A Y A SANHITA, 2023
NO. 45 OF 2023
[25th December ,2023.]
An Act to consolidate and amend the provisions relating to offences and for
matters connected therewith or incidental thereto.
BE it enacted by Parliament in the Seventy-fourth Y ear of the Republic of India as
follows:––
CHAPTER I
PRELIMINARY
1.(1) This Act may be called the Bharatiya Nyaya Sanhita, 2023.
(2) It shall come into force on such date as the Central Government may , by notification
in the Official Gazette, appoint, and different dates may be appointed for different provisions
of this Sanhita.
Short title,
commencement
and
application.
vlk/kkj.k
EXTRAORDINARY
Hkkx II — [k.M 1
P ART II — Section 1
izkf/kdkj ls izdkf'kr
PUBLISHED BY AUTHORITY
lañ 53] ubZ fnYyh] lkseokj] fnlEcj 25] 2023@ ikS"k 4] 1945 ¼'kd½
No. 53] NEW DELHI, MONDAY, DECEMBER 25, 2023/PAUSHA 4, 1945 (SAKA)
bl Hkkx esa fHkUu i`"B la[;k nh tkrh gS ftlls fd ;g vyx ladyu ds :i esa j[kk tk ldsA
Separate paging is given to this Part in orde

# Chunking

In [22]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000,
    chunk_overlap=200
)

chunks = splitter.split_documents(docs)

print("Total Chunks:", len(chunks))

Total Chunks: 534


In [20]:
print(chuncks[0].page_content)

THE BHARA TIY A NY A Y A SANHITA, 2023
NO. 45 OF 2023
[25th December ,2023.]
An Act to consolidate and amend the provisions relating to offences and for
matters connected therewith or incidental thereto.
BE it enacted by Parliament in the Seventy-fourth Y ear of the Republic of India as
follows:––
CHAPTER I
PRELIMINARY
1.(1) This Act may be called the Bharatiya Nyaya Sanhita, 2023.
(2) It shall come into force on such date as the Central Government may , by notification
in the Official Gazette, appoint, and different dates may be appointed for different provisions
of this Sanhita.
Short title,
commencement
and
application.
vlk/kkj.k
EXTRAORDINARY
Hkkx II — [k.M 1
P ART II — Section 1
izkf/kdkj ls izdkf'kr
PUBLISHED BY AUTHORITY
lañ 53] ubZ fnYyh] lkseokj] fnlEcj 25] 2023@ ikS"k 4] 1945 ¼'kd½
No. 53] NEW DELHI, MONDAY, DECEMBER 25, 2023/PAUSHA 4, 1945 (SAKA)
bl Hkkx esa fHkUu i`"B la[;k nh tkrh gS ftlls fd ;g vyx ladyu ds :i esa j[kk tk ldsA
Separate paging is given to this Part in orde

This metadata is extremely important because later, when the RAG answers a question, you can cite the page number from which the information was retrieved.

*after generating this we can proceed to Embedding*

In [24]:
print(chunks[0].metadata)

{'producer': 'cairo 1.17.4 (https://cairographics.org)', 'creator': 'Mozilla Firefox 124.0.1', 'creationdate': '2024-04-01T12:07:18+05:30', 'source': 'CON.pdf', 'total_pages': 102, 'page': 0, 'page_label': '1'}


#Embeddings

In [8]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer(
    "all-MiniLM-L6-v2"
)

vectors = [
    model.encode(chunk.page_content)
    for chunk in chuncks
]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

This is typical extraction of the embeddings, giving the value of the vectors.

In [27]:
!pip install sentence-transformers

from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

chunk_texts = [chunk.page_content for chunk in chunks]

embeddings = model.encode(chunk_texts)

print(type(embedding))
print("Dimensions:", len(embedding))
print("First 10 values:", embedding[:10])

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

<class 'numpy.ndarray'>
Dimensions: 384
First 10 values: [-0.05945378  0.11856074  0.05896226  0.01657     0.06958461  0.13816461
  0.03750493 -0.02968621 -0.03641447  0.03982205]


Output:
Dimensions: 384
First 10 values: [-0.05945378  0.11856074  0.05896226  0.01657     0.06958461  0.13816461
  0.03750493 -0.02968621 -0.03641447  0.03982205]

*Printed the Embedding*
and  embedding.shape defins that sentence into a 384-dimensional vector.

In [26]:
print(embedding)
embedding.shape

[-5.94537780e-02  1.18560739e-01  5.89622632e-02  1.65700037e-02
  6.95846081e-02  1.38164610e-01  3.75049300e-02 -2.96862070e-02
 -3.64144668e-02  3.98220494e-02  9.41409692e-02  1.23510920e-01
  1.81037653e-02 -9.42312367e-03  3.13711613e-02 -1.97518636e-02
 -4.01775800e-02 -8.12199432e-03 -6.44879118e-02  3.02696172e-02
  1.42901549e-02  1.37422672e-02  2.40260642e-02  4.78531830e-02
 -5.99767566e-02 -1.07742315e-02 -5.17689483e-03 -2.45546103e-02
  4.00282256e-02 -1.13305738e-02  1.04880393e-01 -2.94053499e-02
 -4.93795648e-02 -7.35876197e-03 -2.72317920e-02 -3.72044928e-02
  5.44966683e-02 -4.65338156e-02  7.09539745e-03  7.39427563e-03
 -5.00004999e-02 -9.43091884e-02 -4.55826074e-02  5.73968105e-02
 -5.48862561e-04  1.71137452e-02 -3.06123532e-02 -1.35904830e-02
 -9.37873963e-03  1.52925309e-02  1.30216544e-02  2.00710893e-02
 -3.95207703e-02 -2.59520882e-03  4.78976369e-02  6.84045581e-03
 -4.25357595e-02 -4.69074585e-02 -1.65670216e-02 -1.09169278e-02
  2.65645776e-02 -7.24714

(384,)

*NOW we are heading towards the sementic search and Extraction*

In [29]:
!pip install scikit-learn
from sklearn .metrics.pairwise import cosine_similarity
import numpy as np

what we have done here is that we have first created a query and convert this into the embedings

In [31]:
query= "what is private defence?"
query_embedding = model.encode(query)

In [33]:
query_embedding.shape

(384,)

As we have observed till now that <br>
Chunk embedding-> 384<br>
query embedding-> 384<br>

lets  compare this using cosine similarity

In [35]:
scores = cosine_similarity(
    [query_embedding],
    embeddings
)[0]

In [38]:
print(scores[:10])

[0.2631629  0.10556225 0.29786605 0.26693237 0.09024294 0.09936758
 0.13231903 0.0942225  0.18952219 0.19352072]


In [41]:
best_index = np.argmax(scores)

print("Best Index:", best_index)

print(chunks[best_index].page_content)

Best Index: 67
be strictly justifiable by law;
(c) in cases in which there is time to have recourse to the protection of the public
authorities.
(2) The right of private defence in no case extends to the inflicting of more harm than
it is necessary to inflict for the purpose of defence.
Explanation 1.—A person is not deprived of the right of private defence against an act
done, or attempted to be done, by a public servant, as such, unless he knows or has reason
to believe, that the person doing the act is such public servant.
Explanation 2.—A person is not deprived of the right of private defence against an act
done, or attempted to be done, by the direction of a public servant, unless he knows, or has
reason to believe, that the person doing the act is acting by such direction, or unless such
person states the authority under which he acts, or if he has authority in writing, unless he
produces such authority, if demanded.


In [45]:
top_k=5
top_indices= np.argsort(scores) [-top_k:][::-1]

In [46]:
for i, idx in enumerate(top_indices):
    print(f"\nResult {i+1}")
    print("Score:", scores[idx])
    print("Page:", chunks[idx].metadata["page"])
    print(chunks[idx].page_content[:1000])


Result 1
Score: 0.5434712
Page: 14
be strictly justifiable by law;
(c) in cases in which there is time to have recourse to the protection of the public
authorities.
(2) The right of private defence in no case extends to the inflicting of more harm than
it is necessary to inflict for the purpose of defence.
Explanation 1.—A person is not deprived of the right of private defence against an act
done, or attempted to be done, by a public servant, as such, unless he knows or has reason
to believe, that the person doing the act is such public servant.
Explanation 2.—A person is not deprived of the right of private defence against an act
done, or attempted to be done, by the direction of a public servant, unless he knows, or has
reason to believe, that the person doing the act is acting by such direction, or unless such
person states the authority under which he acts, or if he has authority in writing, unless he
produces such authority, if demanded.

Result 2
Score: 0.49707276
Page: 15
(f) a

WE HAVE Created a RAG System for the Indian Constitution using the pdf, now lets create a reusable test function and TEST it !


In [47]:
def search(query, k=5):
    query_embedding = model.encode(query)

    scores = cosine_similarity(
        [query_embedding],
        embeddings
    )[0]

    top_indices = np.argsort(scores)[-k:][::-1]

    results = []

    for idx in top_indices:
        results.append({
            "score": float(scores[idx]),
            "page": chunks[idx].metadata["page"],
            "text": chunks[idx].page_content
        })

    return results

In [48]:
results = search("What is murder?")

for r in results:
    print("="*80)
    print("Score:", r["score"])
    print("Page:", r["page"])
    print(r["text"][:1000])

Score: 0.5883523225784302
Page: 31
a man in the ordinary course of nature. Z dies in consequence. Here A is guilty of murder,
although he may not have intended to cause Z’s death.
(d) A without any excuse fires a loaded cannon into a crowd of persons and kills one
of them. A is guilty of murder, although he may not have had a premeditated design to kill any
particular individual.
Exception 1.—Culpable homicide is not murder if the offender, whilst deprived of the
power of self-control by grave and sudden provocation, causes the death of the person who
gave the provocation or causes the death of any other person by mistake or accident:
Provided that the provocation is not,––
(a) sought or voluntarily provoked by the offender as an excuse for killing or
doing harm to any person;
(b) given by anything done in obedience to the law, or by a public servant in the
lawful exercise of the powers of such public servant;
(c) given by anything done in the lawful exercise of the right of private de

Excelent job pal, you have just created a RAG system and tested along . This is like running RAG chatbot on the google Collab or any other IDE.

#Configuring LLM

In [49]:
!pip install google-generativeai

create a new API KEY from https://aistudio.google.com/app/api-keys and replace with YOUR-Private-API-Key

In [77]:
import google.generativeai  as genai
genai.configure(api_key='Your-Private-API-Key')

In [78]:
model_gemini = genai.GenerativeModel(
    "gemini-2.5-flash"
)

In [79]:
def retrieve_context(query, k=5):
    query_embedding = model.encode(query)

    scores = cosine_similarity(
        [query_embedding],
        embeddings
    )[0]

    top_indices = np.argsort(scores)[-k:][::-1]

    contexts = []

    for idx in top_indices:
        contexts.append(
            chunks[idx].page_content
        )

    return "\n\n".join(contexts)

In [80]:
context = retrieve_context(
    "What is murder?"
)

print(context[:2000])


a man in the ordinary course of nature. Z dies in consequence. Here A is guilty of murder,
although he may not have intended to cause Z’s death.
(d) A without any excuse fires a loaded cannon into a crowd of persons and kills one
of them. A is guilty of murder, although he may not have had a premeditated design to kill any
particular individual.
Exception 1.—Culpable homicide is not murder if the offender, whilst deprived of the
power of self-control by grave and sudden provocation, causes the death of the person who
gave the provocation or causes the death of any other person by mistake or accident:
Provided that the provocation is not,––
(a) sought or voluntarily provoked by the offender as an excuse for killing or
doing harm to any person;
(b) given by anything done in obedience to the law, or by a public servant in the
lawful exercise of the powers of such public servant;
(c) given by anything done in the lawful exercise of the right of private defence.

102.If a person, by doing a

Create a RAG Funcion & TEST

In [81]:
def ask_rag(query):

    context = retrieve_context(query)

    prompt = f"""
You are a legal assistant.

Answer ONLY using the provided context.

Context:
{context}

Question:
{query}

Answer:
"""

    response = model_gemini.generate_content(
        prompt
    )

    return response.text

In [82]:
answer = ask_rag(
    "What is murder?"
)

print(answer)

Except in the cases hereinafter excepted, culpable homicide is murder,––
(a) if the act by which the death is caused is done with the intention of causing death; or
(b) if the act by which the death is caused is done with the intention of causing such bodily injury as the offender knows to be likely to cause the death of the person to whom the harm is caused; or
(c) if the act by which the death is caused is done with the intention of causing bodily injury to any person, and the bodily injury intended to be inflicted is sufficient in the ordinary course of nature to cause death; or
(d) if the person committing the act knows that it is so imminently dangerous that it must, in all probability, cause death, or such bodily injury as is likely to cause death, and commits such act without any excuse for incurring the risk of causing death or such injury as aforesaid.


Ask more query

In [83]:
ask_rag(
    "What is private defence?"
)

'Every person has a right, subject to the restrictions contained in section 37, to defend:\n(a) his own body, and the body of any other person, against any offence affecting the human body;\n(b) the property, whether movable or immovable, of himself or of any other person, against any act of theft, mischief, criminal trespass, or an attempt to commit any such offence.'

Congratulations ! At this time we have implemented & tested the complete RAG System.

Test More queries.
Happy GenAI! :)